bridge-fungible-token/src/main.sw

```diff
@@ -1,69 +1,186 @@
+    identity::Identity,
+    logging::log,
+    option::Option,
+    result::Result,
+    revert::{
+        require,
+        revert,
+    },
+    storage::StorageMap,
+    u256::U256,
+    vm::evm::evm_address::EvmAddress,
+};
+use utils::{
+    burn_tokens,
+    correct_input_type,
+    decompose,
+    input_message_data,
+    input_message_data_length,
+    input_message_recipient,
+    input_message_sender,
+    is_address,
+    mint_tokens,
+    parse_message_data,
+    send_message,
+    transfer_tokens,
+};
 
+////////////////////////////////////////
+// Constants
+////////////////////////////////////////
+const NAME = "PLACEHOLDER";
+const SYMBOL = "PLACEHOLDER";
+const DECIMALS = 9u8;
+const LAYER_1_DECIMALS = 18u8;
+
+////////////////////////////////////////
+// Storage declarations
+////////////////////////////////////////
 storage {
-    counter: u64 = 0,
-    data1: ContractId = ~ContractId::from(ZERO_B256),
-    data2: u64 = 0,
-    data3: b256 = ZERO_B256,
-    data4: Address = ~Address::from(ZERO_B256),
+    refund_amounts: StorageMap<(b256, EvmAddress), U256> = StorageMap {},
 }
 
+////////////////////////////////////////
+// ABI Implementations
+////////////////////////////////////////
 // Implement the process_message function required to be a message receiver
 impl MessageReceiver for Contract {
     #[storage(read, write)]
     fn process_message(msg_idx: u8) {
-        storage.counter = storage.counter + 1;
+        let input_sender = input_message_sender(1);
 
-        // Parse the message data
-        let data_length = input_message_data_length(msg_idx);
-        if (data_length >= 32) {
-            let contract_id: b256 = input_message_data(msg_idx, 0);
-            storage.data1 = ~ContractId::from(contract_id);
-        }
-        if (data_length >= 32 + 8) {
-            let num: u64 = input_message_data(msg_idx, 32);
-            storage.data2 = num;
-        }
-        if (data_length >= 32 + 8 + 32) {
-            let big_num: b256 = input_message_data(msg_idx, 32 + 8);
-            storage.data3 = big_num;
-        }
-        if (data_length >= 32 + 8 + 32 + 32) {
-            let address: b256 = input_message_data(msg_idx, 32 + 8 + 32);
-            storage.data4 = ~Address::from(address);
+        require(input_sender.value == LAYER_1_ERC20_GATEWAY, BridgeFungibleTokenError::UnauthorizedUser);
+
+        let message_data = parse_message_data(msg_idx);
+
+        // @review verify `l1_asset` matches hardcoded LAYER_1_TOKEN value
+        require(message_data.l1_asset == ~EvmAddress::from(LAYER_1_TOKEN), BridgeFungibleTokenError::IncorrectAssetDeposited);
+
+        // @review verify `from` matches hardcoded PREDICATE_ROOT value
+        require(message_data.from == ~Address::from(PREDICATE_ROOT), BridgeFungibleTokenError::UnauthorizedUser);
+
+        // @review verify `to` matches hardcoded value TO value
+        require(message_data.to == ~Address::from(TO), BridgeFungibleTokenError::UnauthorizedUser);
```

Comment:
I'm not sure what this is doing. The 'to' field should just be who gets the coins once they're minted.

Reply:
This is a mistake that I should have already removed. Doing so now!